File: 0023.MergekSortedLists.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        k = len(lists)
        #return self.depart(lists, 0, k-1)
        if k == 0:
            return lists
        while k>1:
            for i in xrange(k/2):
                lists[i] = self.mergeTwoLists(lists[i], lists[k-i-1])
            k = (k+1)/2
        return lists[0]

    def depart(self, lists, start, end):
        if start == end:
            return lists[start]
        if start < end:
            mid = (start+end)/2
            l1 = self.depart(lists, start, mid)
            l2 = self.depart(lists, mid+1, end)
            return self.mergeTwoLists(l1, l2)

    def mergeTwoLists(self, l1, l2):
        p = head = ListNode(0)
        while l1 and l2:
            if l1.val <= l2.val:
                p.next = l1
                l1 = l1.next
            else:
                p.next = l2
                l2 = l2.next
            p = p.next
        if l1:
            p.next = l1
        if l2:
            p.next = l2
        return head.next

File: 0000-0999/0023.MergekSortedLists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// O(N*logK), 4-8ms < 99.75%; O(1), 5.3MB < 100% 
func mergeKLists(lists []*ListNode) *ListNode {
    if len(lists) == 0 {
        return nil
    }
    return divideConquer(lists, 0, len(lists)-1)
    /*
    n := len(lists)
    if n == 0 {
        return nil
    }
    for n > 1 {
        for i := 0; i < n/2; i++ {
            lists[i] = mergeTwoLists(lists[i], lists[n-i-1])
        }
        n = (n+1) / 2
    }
    return lists[0]
    */
}

func divideConquer(lists []*ListNode, left int, right int) *ListNode {
    if left == right {
        return lists[left]
    }
    mid := (left + right)/2
    l1 := divideConquer(lists, left, mid)
    l2 := divideConquer(lists, mid+1, right)
    return mergeTwoLists(l1, l2)
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
    head := &ListNode{}
    p := head
    for l1 != nil && l2 != nil {
        if l1.Val <= l2.Val {
            p.Next = l1
            l1 = l1.Next
        } else {
            p.Next = l2
            l2 = l2.Next
        }
        p = p.Next
    }
    if l1 != nil {
        p.Next = l1
    } else {
        p.Next = l2
    }
    return head.Next
}

// O(N*K), 108ms < 34.41%; O(1), 5.3MB < 100%
// N = sum(len(lists[i])), K = len(lists)
/*
func mergeKLists(lists []*ListNode) *ListNode {
    n := len(lists)
    if n == 0 {
        return nil
    }
    ans := lists[0]
    for i := 1; i < n; i++ {
        ans = mergeTwoLists(ans, lists[i])
    }
    return ans
}
*/